### PR TITLE
[T5 optimization] script fusions and fixes

### PR DESCRIPTION
**Pull request by wangyems**
### Description

1. added script for t5 encoder self attention and t5 decoder self/cross attention fusions. 
2. added simplified layernorm fusion for --external_data_format senario. (otherwise relying on ORT optimizer)
3. added rel_pos_bias shape inference code, modified attention/mha shape inference script.
4. reworked graph_topologic_sort() because the currently implementation is not functioning correctly. also added an option to topo-sort the graph in a deterministic way to let tests pass.

note:
1. the t5-beamsearch export code is slightly modified. specifically, encoder_hidden_states(ehs) is no longer an input to the t5 decoder since the ehs is not actually used in the graph execution. 
2. recent PRs do not add optimizations to t5 on cpu. 
3. the fp32 model(encoder and decoder) for t5-small, t5-base and t5-large can get a parity of e-5 and the corresponding beam search models generate same results as pytorch.
4. fp16(mixed-precision) models, however, get a parity around 3e-2 and some has maximum diff a bit over 3e-2. But the beam search models still generate same results as pytorch (based on limited input data)
5. mt-5 model has a parity issue at the moment, even before any optimization. will investigate later.

### CODE REVIEW

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 1]
```diff
@@ -378,6 +379,16 @@ def _get_shape(self, node, idx):
             assert name in self.initializers_
             return list(self.initializers_[name].dims)
 
+    def _try_get_shape(self, node, idx):
+        name = node.input[idx]
+        if name in self.known_vi_:
+            vi = self.known_vi_[name]
+            return get_shape_from_value_info(vi)
+        else:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:387]: nit: else is not needed

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 2]
```diff
@@ -1495,6 +1507,17 @@ def _infer_ReduceProd(self, node):
             if data is not None:
                 self.sympy_data_[node.output[0]] = sympy_reduce_product(data)
 
+    def _infer_RelativePositionBias(self, node):
+        new_shape = []
+        new_shape.append("1")
+        new_shape.append(str(self._get_sympy_shape(node, 0)[1]))
+        new_shape.append(str(self._try_get_value(node, 1)))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1514]: what if _try_get_value returns None

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 3]
```diff
@@ -2080,29 +2107,61 @@ def _infer_MultiHeadAttention(self, node):
...
+                total_sequence_length = query_shape[1]
+
                 output_dtype = self.known_vi_[node.input[0]].type.tensor_type.elem_type
                 vi = self.known_vi_[node.output[0]]
                 vi.CopyFrom(helper.make_tensor_value_info(node.output[0], output_dtype, output_shape))
 
+            if len(node.output) > 1:
+                batch_size = query_shape[0]
+                num_heads = get_attribute(node, "num_heads")
+                head_size = int(query_shape[2] / num_heads) if len(query_shape) == 3 else query_shape[4]
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:2141]: check that query_shape[2] is int type?

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 4]
```diff
@@ -2080,29 +2107,61 @@ def _infer_MultiHeadAttention(self, node):
...
+                    and isinstance(num_heads, int)
+                    and isinstance(total_sequence_length, int)
+                    and isinstance(head_size, int)
+                ):
+                    present_shape = [batch_size, num_heads, total_sequence_length, head_size]
+                else:
+                    present_shape = [str(batch_size), str(num_heads), str(total_sequence_length), str(head_size)]
+
+                vi = self.known_vi_[node.output[1]]
+                vi.CopyFrom(helper.make_tensor_value_info(vi.name, output_dtype, present_shape))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:2161]: `output_dtype` might be  unbound (not initialized)

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 5]
```diff
@@ -876,66 +876,62 @@ def is_safe_to_fuse_nodes(self, nodes_to_remove, keep_outputs, input_name_to_nod
...
+                    continue
+                input_count = sum(1 for _ in node.input if _)
+                if input_count == 0:
+                    sorted_nodes.append(node)
+                    sorted_node_set.add(node_idx)
+                    for output in node.output:
+                        deps_set.add(output)
+                    continue
+                failed = False
+                for input in node.input:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:911]: Redefining built-in 'input'. How about input_name?

File: onnxruntime/python/tools/transformers/onnx_model_t5.py [hunk 6]
```diff
@@ -29,25 +29,460 @@ def __init__(
...
+            return
+
+        transpose_q, reshape_q, matmul_q = q_nodes
+
+        if matmul_q.input[0] != input_shape_node.input[0]:
+            return
+
+        q_num_heads, q_hidden_size = self.get_num_heads_and_hidden_size(reshape_q)
+
+        if self.static_kv == 1 and past_key is not None:
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model_t5.py:452]: else return?

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 7]
```diff
@@ -876,66 +876,62 @@ def is_safe_to_fuse_nodes(self, nodes_to_remove, keep_outputs, input_name_to_nod
...
-                        if deps_count[node_idx] == 0:
-                            sorted_nodes.append(graph.node[node_idx])
-                            end = end + 1
-            start = start + 1
-
-        if end != len(graph.node):
+        if len(sorted_node_set) != len(graph.node):
             raise RuntimeError(
-                f"Graph is not a DAG: end={end}, len(graph.node)={len(graph.node)}, graph.node[end]={graph.node[end]}"
+                f"Graph is not a DAG: len(sorted_node_set)={len(sorted_node_set)}, len(graph.node)={len(graph.node)}"
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:924]: graph.node[end] is helpful to identify the node that causes non DAG. Please add it to the error message.

File: onnxruntime/python/tools/transformers/onnx_model_t5.py [hunk 8]
```diff
@@ -29,25 +29,460 @@ def __init__(
...
+        attention_node = helper.make_node(
+            "MultiHeadAttention",
+            inputs=attention_inputs,
+            outputs=attention_outputs,
+            name=attention_node_name,
+        )
+
+        attention_node.domain = "com.microsoft"
+        attention_node.attribute.extend([helper.make_attribute("num_heads", num_heads)])
+        attention_node.attribute.extend([helper.make_attribute("scale", 1.0)])
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model_t5.py:101]: why 1.0 here?

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 9]
```diff
@@ -1495,6 +1507,17 @@ def _infer_ReduceProd(self, node):
             if data is not None:
                 self.sympy_data_[node.output[0]] = sympy_reduce_product(data)
 
+    def _infer_RelativePositionBias(self, node):
+        new_shape = []
+        new_shape.append("1")
+        new_shape.append(str(self._get_sympy_shape(node, 0)[1]))
+        new_shape.append(str(self._try_get_value(node, 1)))
+        new_shape.append(str(self._try_get_value(node, 2)))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:1515]: same here

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 10]
```diff
@@ -2030,14 +2053,18 @@ def _infer_ZipMap(self, node):
 
     def _infer_Attention(self, node):
         shape = self._get_shape(node, 0)
-        shape_bias = self._get_shape(node, 2)
-        if shape and len(shape) == 3 and shape_bias and len(shape_bias) == 1:
+        shape_weights = self._get_shape(node, 1)
+        shape_bias = self._try_get_shape(node, 2)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:2057]: what if shape_bias is None?

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 2 above)
Comment (wangyems) [reply onnxruntime/python/tools/symbolic_shape_infer.py:1514]: good catch

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 3 above)
Comment (wangyems) [reply onnxruntime/python/tools/symbolic_shape_infer.py:2141]: fixed

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 11]
```diff
@@ -2080,29 +2107,61 @@ def _infer_MultiHeadAttention(self, node):
...
+
+                if (
+                    isinstance(batch_size, int)
+                    and isinstance(num_heads, int)
+                    and isinstance(total_sequence_length, int)
+                    and isinstance(head_size, int)
+                ):
+                    present_shape = [batch_size, num_heads, total_sequence_length, head_size]
+                else:
+                    present_shape = [str(batch_size), str(num_heads), str(total_sequence_length), str(head_size)]
```
[the remark this answers is not in the document]
Comment (wangyems) [reply onnxruntime/python/tools/symbolic_shape_infer.py:2158]: seems like I misread the helper function. it supports both int and str. fixed

File: onnxruntime/python/tools/transformers/onnx_model_t5.py (hunk 6 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/onnx_model_t5.py:452]: in cross attention the past_* is first assigned to var past_*. since the schema changed the past_* to 'key' and 'value' so we re-assign the input here

File: onnxruntime/python/tools/transformers/onnx_model.py (hunk 7 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/onnx_model.py:924]: added

File: onnxruntime/python/tools/transformers/onnx_model_t5.py (hunk 8 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/onnx_model_t5.py:101]: https://github.com/huggingface/transformers/blob/main/src/transformers/models/t5/modeling_t5.py#L530
In t5 implementation the qk is not scaled

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 10 above)
Comment (wangyems) [reply onnxruntime/python/tools/symbolic_shape_infer.py:2057]: the hidden_size will be read from weights

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 12]
```diff
@@ -2080,29 +2108,67 @@ def _infer_MultiHeadAttention(self, node):
...
+
+                if past_shape is not None:
+                    if isinstance(past_shape[2], int) and isinstance(total_sequence_length, int):
+                        total_sequence_length = past_shape[2] + total_sequence_length
+                    else:
+                        total_sequence_length = f"{past_shape[2]}+{total_sequence_length}"
+
+                # Keep the batch size and total sequence length as string
+                present_shape = [str(batch_size), num_heads, str(total_sequence_length), head_size]
+                print(present_shape)
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:2164]: nit: remove print

File: onnxruntime/python/tools/symbolic_shape_infer.py [hunk 13]
```diff
@@ -2080,29 +2108,67 @@ def _infer_MultiHeadAttention(self, node):
...
+                past_shape = self._try_get_shape(node, 5)
+
+                if past_shape is not None:
+                    if isinstance(past_shape[2], int) and isinstance(total_sequence_length, int):
+                        total_sequence_length = past_shape[2] + total_sequence_length
+                    else:
+                        total_sequence_length = f"{past_shape[2]}+{total_sequence_length}"
+
+                # Keep the batch size and total sequence length as string
+                present_shape = [str(batch_size), num_heads, str(total_sequence_length), head_size]
```
Comment (tianleiwu) [inline onnxruntime/python/tools/symbolic_shape_infer.py:2163]: I think we can remove str() here. Did you encounter issue without explicitly converting to string?

File: onnxruntime/python/tools/transformers/fusion_attention.py [hunk 14]
```diff
@@ -211,6 +212,7 @@ def create_attention_node(
         input: str,
         output: str,
         add_qk_str: str,
+        scale: float = None,
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_attention.py:215]: type is Optional[float]

File: onnxruntime/python/tools/transformers/models/t5/t5_decoder.py [hunk 15]
```diff
@@ -362,7 +350,7 @@ def onnxruntime_inference(ort_session, inputs: T5DecoderInputs):
         ort_inputs = {
             "input_ids": numpy.ascontiguousarray(inputs.decoder_input_ids.cpu().numpy()),
             "encoder_attention_mask": numpy.ascontiguousarray(inputs.encoder_attention_mask.cpu().numpy()),
-            "encoder_hidden_states": numpy.ascontiguousarray(inputs.encoder_hidden_states.cpu().numpy()),
+            # "encoder_hidden_states": numpy.ascontiguousarray(inputs.encoder_hidden_states.cpu().numpy()),
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/t5/t5_decoder.py:353]: remove the commented code.

File: onnxruntime/python/tools/transformers/models/t5/t5_helper.py [hunk 16]
```diff
@@ -151,21 +151,18 @@ def export_onnx(
...
         onnx_model: OnnxModel,
         op_block_list: List[str] = [
-            "Pow",
-            "ReduceMean",
-            "Add",
-            "Sqrt",
-            "Div",
+            "SimplifiedLayerNormalization",
+            "SkipSimplifiedLayerNormalization",
             "Mul",
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/models/t5/t5_helper.py:156]: I think we can remove Mul (which was for mask filter value).

File: onnxruntime/python/tools/transformers/convert_generation.py [hunk 17]
```diff
@@ -723,7 +722,7 @@ def verify_t5_decoder_subgraph(graph: onnx.GraphProto, precision: Precision):
     #   ... (for each cross attention layer)
 
     # TODO: encoder_hidden_states is optional
-    expected_inputs = ["input_ids", "encoder_attention_mask", "encoder_hidden_states"]
+    expected_inputs = ["input_ids", "encoder_attention_mask"]
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/convert_generation.py:725]: If we remove "encoder_hidden_states", we will need update BeamSearch code, which checks the subgraph inputs and outputs.

File: onnxruntime/python/tools/transformers/models/t5/t5_helper.py (hunk 16 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/models/t5/t5_helper.py:156]: removed

File: onnxruntime/python/tools/transformers/convert_generation.py (hunk 17 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/convert_generation.py:725]: the support for decoder without encoder_hidden_states was added before. we can now use it with no change.

File: onnxruntime/python/tools/symbolic_shape_infer.py (hunk 13 above)
Comment (wangyems) [reply onnxruntime/python/tools/symbolic_shape_infer.py]: removed. actually found a bug in retrieving past_key from mha. the index was fixed.

**Review by tianleiwu — APPROVED** [review]

**Review by tianleiwu — APPROVED** [review]
